hibernate/hibernate-ogm
# [PREVIEW] OGM-366 Neo4j remote

https://hibernate.atlassian.net/browse/OGM-366

This pull request creates a new Provider and Dialect for Neo4j remote server.
It is possible to test it using:
```
mvn clean install -DdatastoreProvider=neo4j
``` 
The server needs to be configured with username "neo4j" and password "admin" (these can be changed in the hibernate.properties file). 

As discussed, I removed the ThreadLocal and now the dialect can access the TransactionCoordinator from the various Context (TupleContext, AssociationContext and a new QueryContext).

I wonder if we should continue to support the embedded mode since Neo4j is planning to remove it.
Currently, I've kept the previous dialect but it should be possible to reuse one dialect and one provider (for embedded and remote) by creating some interfaces for the execution of queries and getting the results.

This pull request looks quite big but it is not as big as ti looks like, I've split it in several commits, and the first ones are mainly for preparing the road to the new dialect.

So, the first commits will:
* Make sure that the tests in Neo4j read the default configuration from the property file.
* Add tests to check that the sequence increments is thread safe. There is already one but it only tests for the table generation strategy. Eventually we should add this tests in core and make sure that they are executed for the other db as well.
* upgrade to Neo4j 2.6
* Move the DatabaseIdentifier class  in CouchDB to core, so that it can be reused
* Make sure that the AssociationKey default implementation checks the collectionRole in the equals method (forgot the hashcode, but I will update it)

After all those things are in place the patch will:
* Add the classes to map the requests to the server

  All the operation on the server are executed via Cypher. The Neo4j API allows the execution of multiple queries inside the same request and these queries can be executed inside a running transaction or they can be executed immediately (the server will open a new transaction and it will commit it).

* To execute queries inside an open transaction via Rest I need the the transaction Id. I've changed the various context so that they can contain the TransactionCoordinator so that a dialect can get an Id before any operation.

* Add some new GridType converter that are not need for the embedded use case

* Add the classes for the Remote Neo4j datastore and dialect.

The dialect for the embedded Neo4j uses mainly Cypher queries but some parts are still using the programmatic API (property update, embedded association creation, sequences).

For the dialect I had to convert those parts into Cypher queries and it was a bit more complicated than expected. This resulted in two dialects and providers while in theory we don't need that as we can execute Cypher queries  in both backends.

Another difference of the remote dialect is that it creates a certain amount of queries and then executes them with one request to the server while the embedded can execute a query and update a graph element as soon as needed.

In the end, I think the most complicated part is the generation of the Cypher queries used by the dialects, in particular  the one related to the embedded elements.

Cypher is also missing an operator to acquire a lock an a graph element, reading around I've found a way to do it buy changing a property on a node and than updating the same node inside the same transaction. This seems to work.

This is the gist of it. I hope this help as a high level overview and to start asking for question.
